src/main/scala/io/hydrosphere/serving/gateway/grpc/Reporting.scala

```diff
@@ -0,0 +1,150 @@
+package io.hydrosphere.serving.gateway.grpc
+
+import cats.data.NonEmptyList
+import cats.effect.{Async, IO, LiftIO}
+import cats.syntax.applicative._
+import cats.syntax.applicativeError._
+import cats.syntax.flatMap._
+import cats.syntax.functor._
+import cats.{Applicative, Functor, Monad}
+import io.grpc.Channel
+import io.hydrosphere.serving.gateway.config.Configuration
+import io.hydrosphere.serving.gateway.grpc.reqstore.{Destination, ReqStore}
+import io.hydrosphere.serving.gateway.service.application.ExecutionUnit
+import io.hydrosphere.serving.grpc.AuthorityReplacerInterceptor
+import io.hydrosphere.serving.monitoring.monitoring.ExecutionInformation.ResponseOrError
+import io.hydrosphere.serving.monitoring.monitoring.MonitoringServiceGrpc.MonitoringServiceStub
+import io.hydrosphere.serving.monitoring.monitoring.{ExecutionInformation, ExecutionMetadata, MonitoringServiceGrpc, TraceData}
+import io.hydrosphere.serving.tensorflow.api.predict.PredictRequest
+
+import scala.concurrent.Future
+import scala.concurrent.duration.Duration
+
+trait Reporter[F[_]] {
+  def send(execInfo: ExecutionInformation): F[Unit]
+}
+
+object Reporter {
+
+  def apply[F[_]](f: ExecutionInformation => F[Unit]): Reporter[F] =
+    new Reporter[F] {
+      def send(execInfo: ExecutionInformation): F[Unit] = f(execInfo)
+    }
+
+  def fromFuture[F[_]: Functor, A](f: ExecutionInformation => Future[A])(implicit F: LiftIO[F]): Reporter[F] =
+    Reporter(info => F.liftIO(IO.fromFuture(IO(f(info)))).void)
```

Comment:
Could we use change it to Async + Future.onComplete?

Reply:
If you mean asynchronous sending `analyze` requests (on another execution context) to return serving response faster, then yes. I forgot about that.